src/libcrun/linux.c

```diff
@@ -1963,6 +1963,53 @@ append_mode_if_missing (char *data, const char *mode)
   return new_data;
 }
 
+static int
+safe_create_symlink (int rootfsfd, const char *rootfs, size_t rootfs_len, const char *target, const char *destination, libcrun_error_t *err)
+{
+  cleanup_close int parent_dir_fd = -1;
+  cleanup_free char *buffer = NULL;
+  char *part;
+  int ret;
+
+  if (is_empty_string (destination))
+    return crun_make_error (err, 0, "empty destination for symlink `%s`", target);
+
+  buffer = xstrdup (destination);
+  part = dirname (buffer);
+
+  parent_dir_fd = crun_safe_create_and_open_ref_at (true, rootfsfd, rootfs, rootfs_len,
+                                                    part, 0755, err);
+  if (UNLIKELY (parent_dir_fd < 0))
+    return crun_make_error (err, errno, "symlink creation");
+
+  /* It is safe to reuse the buffer since it was created with xstrdup (destination).  */
+  strcpy (buffer, destination);
+  part = basename (buffer);
+
+  ret = symlinkat (target, parent_dir_fd, part);
+  if (UNLIKELY (ret < 0))
+    {
+      /* If it exists, check if it has the same content, if so just ignore the error.  */
+      if (errno == EEXIST)
+        {
+          cleanup_free char *link = NULL;
+          ssize_t len;
+
+          len = safe_readlinkat (parent_dir_fd, part, &link, 0, err);
+          if (UNLIKELY (len < 0))
+            return len;
+
+          if (strlen (target) == len && strncmp (link, target, len) == 0)
+            return 0;
+
+          return crun_make_error (err, 0, "symlink `%s` already exists with a different content", destination);
+        }
```

Comment:
I have no context of use-case so  I am not sure but should this just override the symlink if it already exists instead of erroring out or atleast the docs be updated.

Reply:
I think it is better to not try to override the destination.  Currently we error out if we try to mount a directory on a file, and the other way around, runc does the same thing.  Also, "replacing" can be expensive since the destination can be a directory.

```
$ touch /tmp/foo; podman run -v /tmp/foo:/tmp --rm -ti alpine sh
Error: OCI runtime error: crun: mount `/tmp/foo` to `tmp`: Not a directory
```